### PR TITLE
Compile plugins defined in a package when building the package

### PR DESCRIPTION
**Pull request by abertelrud**
This improves package authoring by surfacing errors before having to run a command.

Incremental builds don't load the package graph if the package structure and manifest contents haven't changed, so to make this happen we need to also store basic information about the plugins in the BuildDescription.  We don't want to incorporate these commands into the build graph at this point, since it would be a much bigger change and also complicate the compile-on-demand behavior of the plugins, plus require all clients of libSwiftPM to add a dependency on llbuild — but it's possible that we'll want to make that restructuring in the future (in particular if it would allow a plugin to use a library in an easier way).

Specific changes in this PR:

- move PluginCompilationResult from DefaultPluginScriptRunner to PluginInvocation so it can be referenced by the PluginScriptRunner protocol
- add a method to PluginScriptRunner to compile but not run a plugin and just return a compilation result
- add a PluginDescription to BuildPlan, and serialize it into the BuildDescription so we don't have to reload the graph
- add a PlugInScriptRunner to BuildOperation, and have SwiftTool (and specializations of it) pass this down to the operation when creating it
- replace the `buildToolPluginInvoker` closure with a PluginScriptRunner and pluginWorkingDirectory pair (a bit less of a layering violation)
- extend BuildOperationBuildSystemDelegateHandler to also be called when a preparation step starts, emits output, and finishes
- compile the plugins in the build set before starting the build
- add a unit test that tests the new behavior specifically:
   - building compiles all plugins in the root package and then proceeds with build
   - but if one of them fails it does not proceed with the build
   - naming a specific plugin using `--target` builds just that plugin but doesn't build anything else

As a bonus due to some of the underlying work this also makes build tool plugins available in all build operations, including the API Digester etc.

rdar://88084413

### CODE REVIEW

**Comment by abertelrud** [comment]
@swift-ci please smoke test

File: Sources/Build/BuildOperation.swift [hunk 1]
```diff
@@ -29,9 +29,12 @@ public final class BuildOperation: PackageStructureDelegate, SPMBuildCore.BuildS
     /// The closure for loading the package graph.
     let packageGraphLoader: () throws -> PackageGraph
     
-    /// The closure for invoking build tool plugins in the package graph.
-    let buildToolPluginInvoker: (PackageGraph) throws -> [ResolvedTarget: [BuildToolPluginInvocationResult]]
```
Comment (abertelrud) [inline Sources/Build/BuildOperation.swift:33]: I'm particularly happy about getting rid of this.  Although it mirrored the `packageGraphLoader`, it always felt kind of dodgy.

File: Sources/Build/BuildOperationBuildSystemDelegateHandler.swift [hunk 2]
```diff
@@ -594,6 +598,22 @@ final class BuildOperationBuildSystemDelegateHandler: LLBuildBuildSystemDelegate
     func shouldResolveCycle(rules: [BuildKey], candidate: BuildKey, action: CycleAction) -> Bool {
         return false
     }
+    
+    func preparationStepStarted(_ name: String) {
```
Comment (abertelrud) [inline Sources/Build/BuildOperationBuildSystemDelegateHandler.swift:602]: Not in love with this but I think this is conceptually correct as long as these steps happen outside the build system — I tried impersonating build commands, but it made things harder to understand, and so this is at least clear.  We should consider refactoring the output emission here and should revisit this at that time.  For now this seems the most correct.

Comment (abertelrud) [reply Sources/Build/BuildOperationBuildSystemDelegateHandler.swift:602]: (this comment applies to these three methods, calling out build preparation steps as distinct from the build — in fact I think there are other steps that could fall into this category, to the extent SwiftPM generates modulemaps or does other things that could take time and/or fail)

**Comment by abertelrud** [comment]
@swift-ci please smoke test

**Review by tomerd — APPROVED** [review]
nice!

**Comment by abertelrud** [comment]
Odd, the `@main` support hasn't failed for a long time:
```
20:09:13 Test Case 'BuildToolTests.testAtMainSupport' started at 2022-02-03 04:07:47.438
20:09:13 
/home/build-user/swiftpm/Tests/CommandsTests/BuildToolTests.swift:183: error: BuildToolTests.testAtMainSupport : XCTAssertTrue failed - unexpected failure matching '["SwiftExecSingleFile.product", "SwiftExecMultiFile.product", "description.json", "ClangExecSingleFile.product"]' against pattern [TSCTestSupport.StringPattern.equal("ClangExecSingleFile")]
20:09:13 /home/build-user/swiftpm/Tests/CommandsTests/BuildToolTests.swift:190: error: BuildToolTests.testAtMainSupport : XCTAssertTrue failed - unexpected failure matching '["SwiftExecSingleFile.product", "SwiftExecMultiFile.product", "description.json", "ClangExecSingleFile.product"]' against pattern [TSCTestSupport.StringPattern.equal("SwiftExecSingleFile")]
20:09:13 /home/build-user/swiftpm/Tests/CommandsTests/BuildToolTests.swift:197: error: BuildToolTests.testAtMainSupport : XCTAssertTrue failed - unexpected failure matching '["SwiftExecSingleFile.product", "SwiftExecMultiFile.product", "description.json", "ClangExecSingleFile.product"]' against pattern [TSCTestSupport.StringPattern.equal("SwiftExecMultiFile")]
20:09:13 Test Case 'BuildToolTests.testAtMainSupport' failed (6.059 seconds)
```

**Comment by abertelrud** [comment]
> Odd, the `@main` support hasn't failed for a long time:
> 
> ```
> 20:09:13 Test Case 'BuildToolTests.testAtMainSupport' started at 2022-02-03 04:07:47.438
> 20:09:13 
> /home/build-user/swiftpm/Tests/CommandsTests/BuildToolTests.swift:183: error: BuildToolTests.testAtMainSupport : XCTAssertTrue failed - unexpected failure matching '["SwiftExecSingleFile.product", "SwiftExecMultiFile.product", "description.json", "ClangExecSingleFile.product"]' against pattern [TSCTestSupport.StringPattern.equal("ClangExecSingleFile")]
> 20:09:13 /home/build-user/swiftpm/Tests/CommandsTests/BuildToolTests.swift:190: error: BuildToolTests.testAtMainSupport : XCTAssertTrue failed - unexpected failure matching '["SwiftExecSingleFile.product", "SwiftExecMultiFile.product", "description.json", "ClangExecSingleFile.product"]' against pattern [TSCTestSupport.StringPattern.equal("SwiftExecSingleFile")]
> 20:09:13 /home/build-user/swiftpm/Tests/CommandsTests/BuildToolTests.swift:197: error: BuildToolTests.testAtMainSupport : XCTAssertTrue failed - unexpected failure matching '["SwiftExecSingleFile.product", "SwiftExecMultiFile.product", "description.json", "ClangExecSingleFile.product"]' against pattern [TSCTestSupport.StringPattern.equal("SwiftExecMultiFile")]
> 20:09:13 Test Case 'BuildToolTests.testAtMainSupport' failed (6.059 seconds)
> ```
> 
> I don't think I saw this locally but it might be Linux-specific. Checking.

No, I just didn't push all my changes properly.

**Comment by abertelrud** [comment]
@swift-ci please smoke test

**Pull request merged**